components/tikv_util/src/stream.rs

```diff
@@ -130,10 +143,98 @@ impl<E> Default for RetryExt<E> {
     fn default() -> Self {
         Self {
             on_failure: Default::default(),
+            max_retry_times: MAX_RETRY_TIMES,
+            max_retry_delay: MAX_RETRY_DELAY,
         }
     }
 }
 
+#[doc(hidden)]
+pub mod __macro_helper {
+    #[doc(hidden)]
+    pub use rand::thread_rng as __thread_rng;
+    #[doc(hidden)]
+    pub use rand::Rng as __rand_Rng;
+    #[doc(hidden)]
+    pub use tokio::time::sleep as __tokio_sleep;
+}
+
+/// JustRetry wraps an error to [`RetryError`] which will always be retryed.
+#[derive(Debug, derive_more::Deref, derive_more::DerefMut, derive_more::Display)]
+pub struct JustRetry<E>(pub E);
+
+impl<E> RetryError for JustRetry<E> {
+    fn is_retryable(&self) -> bool {
+        true
+    }
+}
+
+/// retry_expr! make a future that evaluates the expression and retry when it
+/// evaluated to an `Err`. The expression will be evaluated multi-times.
+///
+/// This would be useful when your action cannot be sealed into a `FnMut`, say,
+/// the action captures some local variables, rustc will complain that `FnMut`
+/// cannot return reference to things it captures.
+///
+/// When possible, prefer using `retry_ext`, which is a normal function. Normal
+/// functions are more friendly to static analysis.
+#[macro_export]
```

Comment:
very cool macro! I had forked the `proc-macro-workshop` long time ago but never had chance to practice on it.

However, not sure if we need to invent wheel by ourselves here, looks like `Tokyo_retry` can do the same thing with better expo backoff option.

example
```
tokio-retry = "0.3"

---
let retries = ExponentialBackoff::from_millis(10)
    .map(10) 
    .take(3); 

let async_fun =  async {
        let data = Cursor::new(data.clone());
        let req = Object::insert_multipart(
            &bucket,
            data,
            content_length,
            &metadata,
            Some(InsertObjectOptional {
                predefined_acl: self.config.predefined_acl,
                ..Default::default()
            }),
        )
        .map_err(RequestError::Gcs)?
        .map(|reader| Body::wrap_stream(AsyncReadAsSyncStreamOfBytes::new(reader)));
        self.make_request(req, tame_gcs::Scopes::ReadWrite).await
    };
Retry::spawn(retries, async_fun).await?;
```

Reply:
`tokio-retry` requires use provide a `FnMut` that makes a future(like `retry_ext` does), when combining with borrowing to local variables, borrow checker will start to complain. Say, when we try to retry some logic that borrows mutable local variables:
```rust
fn somewhat() {
    let mut my_cool_stuff = ();
    let not_retryable = || {
        let borrowed = &mut my_cool_stuff;
        async {
            std::mem::forget(borrowed);
        }
    };
}
```
This code won't compile and rustc complains:
```
  = note: `FnMut` closures only have access to their captured variables while they are executing...
  = note: ...therefore, they cannot allow references to captured variables to escape
```